File: backend/app/api/files.py
"""File Manager API endpoints for browsing, editing, and managing files."""

from flask import Blueprint, request, jsonify, send_file
from flask_jwt_extended import jwt_required
from ..services.file_service import FileService
import os
import tempfile

files_bp = Blueprint('files', __name__)


@files_bp.route('/browse', methods=['GET'])
@jwt_required()
def browse_directory():
    """List directory contents."""
    path = request.args.get('path', '/home')
    show_hidden = request.args.get('show_hidden', 'false').lower() == 'true'

    result = FileService.list_directory(path, show_hidden=show_hidden)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@files_bp.route('/info', methods=['GET'])
@jwt_required()
def get_file_info():
    """Get information about a file or directory."""
    path = request.args.get('path')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    if not FileService.is_path_allowed(path):
        return jsonify({'error': 'Access denied'}), 403

    info = FileService.get_file_info(path)

    if info and 'error' not in info:
        return jsonify({'success': True, 'file': info}), 200

    error = info.get('error', 'File not found') if info else 'File not found'
    return jsonify({'error': error}), 404


@files_bp.route('/read', methods=['GET'])
@jwt_required()
def read_file():
    """Read file contents."""
    path = request.args.get('path')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    result = FileService.read_file(path)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/write', methods=['POST'])
@jwt_required()
def write_file():
    """Write content to a file."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    path = data.get('path')
    content = data.get('content')
    create_backup = data.get('create_backup', True)

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    if content is None:
        return jsonify({'error': 'Content is required'}), 400

    result = FileService.write_file(path, content, create_backup=create_backup)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/create', methods=['POST'])
@jwt_required()
def create_file():
    """Create a new file."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    path = data.get('path')
    content = data.get('content', '')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    result = FileService.create_file(path, content)

    if result.get('success'):
        return jsonify(result), 201

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/mkdir', methods=['POST'])
@jwt_required()
def create_directory():
    """Create a new directory."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    path = data.get('path')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    result = FileService.create_directory(path)

    if result.get('success'):
        return jsonify(result), 201

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/delete', methods=['DELETE'])
@jwt_required()
def delete_path():
    """Delete a file or directory."""
    path = request.args.get('path')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    result = FileService.delete(path)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/rename', methods=['POST'])
@jwt_required()
def rename_path():
    """Rename a file or directory."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    path = data.get('path')
    new_name = data.get('new_name')

    if not path or not new_name:
        return jsonify({'error': 'Path and new_name are required'}), 400

    result = FileService.rename(path, new_name)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/copy', methods=['POST'])
@jwt_required()
def copy_path():
    """Copy a file or directory."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    src = data.get('src')
    dest = data.get('dest')

    if not src or not dest:
        return jsonify({'error': 'Source and destination paths are required'}), 400

    result = FileService.copy(src, dest)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/move', methods=['POST'])
@jwt_required()
def move_path():
    """Move a file or directory."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    src = data.get('src')
    dest = data.get('dest')

    if not src or not dest:
        return jsonify({'error': 'Source and destination paths are required'}), 400

    result = FileService.move(src, dest)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/chmod', methods=['POST'])
@jwt_required()
def change_permissions():
    """Change file/directory permissions."""
    data = request.get_json()

    if not data:
        return jsonify({'error': 'Request body is required'}), 400

    path = data.get('path')
    mode = data.get('mode')

    if not path or not mode:
        return jsonify({'error': 'Path and mode are required'}), 400

    result = FileService.change_permissions(path, mode)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/search', methods=['GET'])
@jwt_required()
def search_files():
    """Search for files matching a pattern."""
    directory = request.args.get('directory', '/home')
    pattern = request.args.get('pattern')
    max_results = request.args.get('max_results', 100, type=int)

    if not pattern:
        return jsonify({'error': 'Search pattern is required'}), 400

    result = FileService.search(directory, pattern, max_results=max_results)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/disk-usage', methods=['GET'])
@jwt_required()
def get_disk_usage():
    """Get disk usage for a path."""
    path = request.args.get('path', '/')

    result = FileService.get_disk_usage(path)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@files_bp.route('/disk-mounts', methods=['GET'])
@jwt_required()
def get_disk_mounts():
    """Get disk usage for all mount points."""
    result = FileService.get_all_disk_mounts()

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@files_bp.route('/analyze', methods=['GET'])
@jwt_required()
def analyze_directory():
    """Analyze directory sizes."""
    path = request.args.get('path', '/home')
    depth = request.args.get('depth', 2, type=int)
    limit = request.args.get('limit', 20, type=int)

    result = FileService.analyze_directory_sizes(path, depth=depth, limit=limit)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/type-breakdown', methods=['GET'])
@jwt_required()
def get_type_breakdown():
    """Get file type breakdown for a directory."""
    path = request.args.get('path', '/home')
    max_depth = request.args.get('max_depth', 3, type=int)

    result = FileService.get_file_type_breakdown(path, max_depth=max_depth)

    if result.get('success'):
        return jsonify(result), 200

    status = 403 if 'denied' in result.get('error', '').lower() else 400
    return jsonify(result), status


@files_bp.route('/download', methods=['GET'])
@jwt_required()
def download_file():
    """Download a file."""
    path = request.args.get('path')

    if not path:
        return jsonify({'error': 'Path is required'}), 400

    if not FileService.is_path_allowed(path):
        return jsonify({'error': 'Access denied'}), 403

    if not os.path.exists(path):
        return jsonify({'error': 'File not found'}), 404

    if os.path.isdir(path):
        return jsonify({'error': 'Cannot download directory'}), 400

    try:
        return send_file(
            path,
            as_attachment=True,
            download_name=os.path.basename(path)
        )
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@files_bp.route('/upload', methods=['POST'])
@jwt_required()
def upload_file():
    """Upload a file."""
    if 'file' not in request.files:
        return jsonify({'error': 'No file provided'}), 400

    file = request.files['file']
    destination = request.form.get('destination')

    if not destination:
        return jsonify({'error': 'Destination path is required'}), 400

    if not FileService.is_path_allowed(destination):
        return jsonify({'error': 'Access denied'}), 403

    if file.filename == '':
        return jsonify({'error': 'No file selected'}), 400

    # Check file size
    file.seek(0, 2)
    size = file.tell()
    file.seek(0)

    if size > FileService.MAX_UPLOAD_SIZE:
        return jsonify({
            'error': f'File too large. Maximum size is {FileService._format_size(FileService.MAX_UPLOAD_SIZE)}'
        }), 400

    try:
        # Determine full path
        if os.path.isdir(destination):
            full_path = os.path.join(destination, file.filename)
        else:
            full_path = destination

        if not FileService.is_path_allowed(full_path):
            return jsonify({'error': 'Access denied'}), 403

        # Ensure parent directory exists
        parent = os.path.dirname(full_path)
        if not os.path.exists(parent):
            os.makedirs(parent)

        # Save file
        file.save(full_path)

        return jsonify({
            'success': True,
            'path': full_path,
            'size': size
        }), 201

    except PermissionError:
        return jsonify({'error': 'Permission denied'}), 403
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: frontend/src/services/api/files.js
// File manager, FTP, Git server, domains, DNS zones, status pages

// File Manager endpoints
export async function browseFiles(path = '/home', showHidden = false) {
    const params = new URLSearchParams({ path, show_hidden: showHidden });
    return this.request(`/files/browse?${params}`);
}

export async function getFileInfo(path) {
    return this.request(`/files/info?path=${encodeURIComponent(path)}`);
}

export async function readFile(path) {
    return this.request(`/files/read?path=${encodeURIComponent(path)}`);
}

export async function writeFile(path, content, createBackup = true) {
    return this.request('/files/write', {
        method: 'POST',
        body: { path, content, create_backup: createBackup }
    });
}

export async function createFile(path, content = '') {
    return this.request('/files/create', {
        method: 'POST',
        body: { path, content }
    });
}

export async function createDirectory(path) {
    return this.request('/files/mkdir', {
        method: 'POST',
        body: { path }
    });
}

export async function deleteFile(path) {
    return this.request(`/files/delete?path=${encodeURIComponent(path)}`, {
        method: 'DELETE'
    });
}

export async function renameFile(path, newName) {
    return this.request('/files/rename', {
        method: 'POST',
        body: { path, new_name: newName }
    });
}

export async function copyFile(src, dest) {
    return this.request('/files/copy', {
        method: 'POST',
        body: { src, dest }
    });
}

export async function moveFile(src, dest) {
    return this.request('/files/move', {
        method: 'POST',
        body: { src, dest }
    });
}

export async function changeFilePermissions(path, mode) {
    return this.request('/files/chmod', {
        method: 'POST',
        body: { path, mode }
    });
}

export async function searchFiles(directory, pattern, maxResults = 100) {
    const params = new URLSearchParams({ directory, pattern, max_results: maxResults });
    return this.request(`/files/search?${params}`);
}

export async function getDiskUsage(path = '/') {
    return this.request(`/files/disk-usage?path=${encodeURIComponent(path)}`);
}

export async function getAllDiskMounts() {
    return this.request('/files/disk-mounts');
}

export async function analyzeDirectory(path = '/home', depth = 2, limit = 20) {
    const params = new URLSearchParams({ path, depth, limit });
    return this.request(`/files/analyze?${params}`);
}

export async function getFileTypeBreakdown(path = '/home', maxDepth = 3) {
    const params = new URLSearchParams({ path, max_depth: maxDepth });
    return this.request(`/files/type-breakdown?${params}`);
}

export async function downloadFile(path) {
    const token = this.getToken();
    const url = `${this.baseUrl}/files/download?path=${encodeURIComponent(path)}`;
    window.open(`${url}&token=${token}`, '_blank');
}

export async function uploadFile(destination, file, onProgress = null) {
    const token = this.getToken();
    const formData = new FormData();
    formData.append('file', file);
    formData.append('destination', destination);

    return new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.open('POST', `${this.baseUrl}/files/upload`);
        xhr.setRequestHeader('Authorization', `Bearer ${token}`);

        if (onProgress) {
            xhr.upload.onprogress = (e) => {
                if (e.lengthComputable) {
                    onProgress((e.loaded / e.total) * 100);
                }
            };
        }

        xhr.onload = () => {
            if (xhr.status >= 200 && xhr.status < 300) {
                resolve(JSON.parse(xhr.responseText));
            } else {
                reject(new Error(JSON.parse(xhr.responseText).error || 'Upload failed'));
            }
        };

        xhr.onerror = () => reject(new Error('Upload failed'));
        xhr.send(formData);
    });
}

// FTP Server endpoints
export async function getFTPStatus() {
    return this.request('/ftp/status');
}

export async function controlFTPService(action, service = null) {
    return this.request(`/ftp/service/${action}`, {
        method: 'POST',
        body: service ? { service } : {}
    });
}

export async function getFTPConfig(service = null) {
    const params = service ? `?service=${service}` : '';
    return this.request(`/ftp/config${params}`);
}

export async function updateFTPConfig(config, service = null) {
    return this.request('/ftp/config', {
        method: 'POST',
        body: { config, service }
    });
}

export async function getFTPUsers() {
    return this.request('/ftp/users');
}

export async function createFTPUser(username, password = null, homeDir = null) {
    return this.request('/ftp/users', {
        method: 'POST',
        body: { username, password, home_dir: homeDir }
    });
}

export async function deleteFTPUser(username, deleteHome = false) {
    const params = deleteHome ? '?delete_home=true' : '';
    return this.request(`/ftp/users/${username}${params}`, {
        method: 'DELETE'
    });
}

export async function changeFTPPassword(username, password = null) {
    return this.request(`/ftp/users/${username}/password`, {
        method: 'POST',
        body: password ? { password } : {}
    });
}

export async function toggleFTPUser(username, enabled) {
    return this.request(`/ftp/users/${username}/toggle`, {
        method: 'POST',
        body: { enabled }
    });
}

export async function getFTPConnections() {
    return this.request('/ftp/connections');
}

export async function disconnectFTPSession(pid) {
    return this.request(`/ftp/connections/${pid}`, {
        method: 'DELETE'
    });
}

export async function getFTPLogs(lines = 100) {
    return this.request(`/ftp/logs?lines=${lines}`);
}

export async function installFTPServer(service = 'vsftpd') {
    return this.request('/ftp/install', {
        method: 'POST',
        body: { service }
    });
}

export async function testFTPConnection(host = 'localhost', port = 21, username = null, password = null) {
    return this.request('/ftp/test', {
        method: 'POST',
        body: { host, port, username, password }
    });
}

// Git Server endpoints
export async function getGitServerStatus() {
    return this.request('/git/status');
}

export async function getGitRequirements() {
    return this.request('/git/requirements');
}

export async function installGit(data) {
    return this.request('/git/install', {
        method: 'POST',
        body: data
    });
}

export async function uninstallGit(removeData = false) {
    return this.request('/git/uninstall', {
        method: 'POST',
        body: { removeData }
    });
}

export async function startGit() {
    return this.request('/git/start', { method: 'POST' });
}

export async function stopGit() {
    return this.request('/git/stop', { method: 'POST' });
}

export async function restartGit() {
    return this.request('/git/restart', { method: 'POST' });
}

// Git Webhooks
export async function getWebhooks() {
    return this.request('/git/webhooks');
}

export async function getWebhook(webhookId) {
    return this.request(`/git/webhooks/${webhookId}`);
}

export async function createWebhook(data) {
    return this.request('/git/webhooks', {
        method: 'POST',
        body: data
    });
}

export async function updateWebhook(webhookId, data) {
    return this.request(`/git/webhooks/${webhookId}`, {
        method: 'PUT',
        body: data
    });
}

export async function deleteWebhook(webhookId) {
    return this.request(`/git/webhooks/${webhookId}`, { method: 'DELETE' });
}

export async function toggleWebhook(webhookId) {
    return this.request(`/git/webhooks/${webhookId}/toggle`, { method: 'POST' });
}

export async function getGitWebhookLogs(webhookId, limit = 50) {
    return this.request(`/git/webhooks/${webhookId}/logs?limit=${limit}`);
}

export async function testWebhook(webhookId) {
    return this.request(`/git/webhooks/${webhookId}/test`, { method: 'POST' });
}

// Git Repositories
export async function getRepositories(limit = 50) {
    return this.request(`/git/repos?limit=${limit}`);
}

export async function getRepository(owner, repo) {
    return this.request(`/git/repos/${owner}/${repo}`);
}

export async function getRepoStats(owner, repo) {
    return this.request(`/git/repos/${owner}/${repo}/stats`);
}

export async function getBranches(owner, repo) {
    return this.request(`/git/repos/${owner}/${repo}/branches`);
}

export async function getBranch(owner, repo, branch) {
    return this.request(`/git/repos/${owner}/${repo}/branches/${branch}`);
}

export async function getCommits(owner, repo, branch = null, page = 1, limit = 30) {
    let url = `/git/repos/${owner}/${repo}/commits?page=${page}&limit=${limit}`;
    if (branch) url += `&branch=${branch}`;
    return this.request(url);
}

export async function getCommit(owner, repo, sha) {
    return this.request(`/git/repos/${owner}/${repo}/commits/${sha}`);
}

export async function getRepoFiles(owner, repo, ref = 'main', path = '') {
    let url = `/git/repos/${owner}/${repo}/contents?ref=${ref}`;
    if (path) url += `&path=${path}`;
    return this.request(url);
}

export async function getFileContent(owner, repo, filepath, ref = 'main') {
    return this.request(`/git/repos/${owner}/${repo}/contents/${filepath}?ref=${ref}`);
}

export async function getRepoReadme(owner, repo, ref = null) {
    let url = `/git/repos/${owner}/${repo}/readme`;
    if (ref) url += `?ref=${ref}`;
    return this.request(url);
}

export async function getGiteaVersion() {
    return this.request('/git/version');
}

// Git Deployments
export async function getAppDeployments(appId, limit = 20) {
    return this.request(`/git/deployments/app/${appId}?limit=${limit}`);
}

export async function getDeployment(deploymentId, includeLogs = false) {
    return this.request(`/git/deployments/${deploymentId}?logs=${includeLogs}`);
}

export async function triggerGitDeploy(appId, branch = null) {
    return this.request(`/git/deployments/app/${appId}/deploy`, {
        method: 'POST',
        body: branch ? { branch } : {}
    });
}

export async function rollbackDeployment(appId, targetVersion = null) {
    return this.request(`/git/deployments/app/${appId}/rollback`, {
        method: 'POST',
        body: targetVersion ? { targetVersion } : {}
    });
}

export async function getWebhookDeployments(webhookId, limit = 20) {
    return this.request(`/git/deployments/webhook/${webhookId}?limit=${limit}`);
}

// Domains endpoints
export async function getDomains() {
    return this.request('/domains');
}

export async function getDomain(id) {
    return this.request(`/domains/${id}`);
}

export async function createDomain(domainData) {
    return this.request('/domains', {
        method: 'POST',
        body: domainData,
    });
}

export async function updateDomain(id, domainData) {
    return this.request(`/domains/${id}`, {
        method: 'PUT',
        body: domainData,
    });
}

export async function deleteDomain(id) {
    return this.request(`/domains/${id}`, {
        method: 'DELETE',
    });
}

export async function enableSsl(domainId, email) {
    return this.request(`/domains/${domainId}/ssl/enable`, {
        method: 'POST',
        body: { email }
    });
}

export async function disableSsl(domainId) {
    return this.request(`/domains/${domainId}/ssl/disable`, { method: 'POST' });
}

export async function renewDomainSsl(domainId) {
    return this.request(`/domains/${domainId}/ssl/renew`, { method: 'POST' });
}

export async function verifyDomain(domainId) {
    return this.request(`/domains/${domainId}/verify`);
}

export async function getDomainsNginxSites() {
    return this.request('/domains/nginx/sites');
}

export async function getDomainsSslStatus() {
    return this.request('/domains/ssl/status');
}

// DNS Zones endpoints
export async function getDNSZones() {
    return this.request('/dns/');
}

export async function getDNSZone(id) {
    return this.request(`/dns/${id}`);
}

export async function createDNSZone(data) {
    return this.request('/dns/', { method: 'POST', body: data });
}

export async function deleteDNSZone(id) {
    return this.request(`/dns/${id}`, { method: 'DELETE' });
}

export async function getDNSRecords(zoneId) {
    return this.request(`/dns/${zoneId}/records`);
}

export async function createDNSRecord(zoneId, data) {
    return this.request(`/dns/${zoneId}/records`, { method: 'POST', body: data });
}

export async function updateDNSRecord(recordId, data) {
    return this.request(`/dns/records/${recordId}`, { method: 'PUT', body: data });
}

export async function deleteDNSRecord(recordId) {
    return this.request(`/dns/records/${recordId}`, { method: 'DELETE' });
}

export async function getDNSPresets() {
    return this.request('/dns/presets');
}

export async function applyDNSPreset(zoneId, preset, variables) {
    return this.request(`/dns/${zoneId}/apply-preset`, {
        method: 'POST', body: { preset, variables }
    });
}

export async function checkDNSPropagation(domain, type = 'A') {
    return this.request(`/dns/propagation/${domain}?type=${type}`);
}

export async function exportDNSZone(zoneId) {
    return this.request(`/dns/${zoneId}/export`);
}

export async function importDNSZone(zoneId, zoneFile) {
    return this.request(`/dns/${zoneId}/import`, {
        method: 'POST', body: { zone_file: zoneFile }
    });
}

// Status Pages endpoints
export async function getStatusPages() {
    return this.request('/status/');
}

export async function getStatusPage(id) {
    return this.request(`/status/${id}`);
}

export async function createStatusPage(data) {
    return this.request('/status/', { method: 'POST', body: data });
}

export async function updateStatusPage(id, data) {
    return this.request(`/status/${id}`, { method: 'PUT', body: data });
}

export async function deleteStatusPage(id) {
    return this.request(`/status/${id}`, { method: 'DELETE' });
}

export async function getPublicStatusPage(slug) {
    return this.request(`/status/public/${slug}`);
}

export async function getStatusPageComponents(pageId) {
    return this.request(`/status/${pageId}/components`);
}

export async function createStatusComponent(pageId, data) {
    return this.request(`/status/${pageId}/components`, { method: 'POST', body: data });
}

export async function updateStatusComponent(compId, data) {
    return this.request(`/status/components/${compId}`, { method: 'PUT', body: data });
}

export async function deleteStatusComponent(compId) {
    return this.request(`/status/components/${compId}`, { method: 'DELETE' });
}

export async function runStatusCheck(compId) {
    return this.request(`/status/components/${compId}/check`, { method: 'POST' });
}

export async function getStatusCheckHistory(compId, hours = 24) {
    return this.request(`/status/components/${compId}/history?hours=${hours}`);
}

export async function getStatusPageIncidents(pageId) {
    return this.request(`/status/${pageId}/incidents`);
}

export async function createStatusIncident(pageId, data) {
    return this.request(`/status/${pageId}/incidents`, { method: 'POST', body: data });
}

export async function updateStatusIncident(incidentId, data) {
    return this.request(`/status/incidents/${incidentId}`, { method: 'PUT', body: data });
}

export async function deleteStatusIncident(incidentId) {
    return this.request(`/status/incidents/${incidentId}`, { method: 'DELETE' });
}

export async function getStatusBadge(slug) {
    return this.request(`/status/badge/${slug}`);
}
